node/src/main/scala/com/wavesplatform/consensus/PoSSelector.scala

```diff
@@ -35,43 +33,60 @@ class PoSSelector(blockchain: Blockchain, blockchainSettings: BlockchainSettings
+        if (vrfActivated(height + 1))
+          getHitSource(height)
+            .map(hs => NxtLikeConsensusBlockData(bt, crypto.signVRF(account.privateKey, hs)))
+        else
+          blockchain.lastBlock
+            .map(_.header.generationSignature.arr)
+            .map(gs => NxtLikeConsensusBlockData(bt, ByteStr(generationSignature(gs, account.publicKey))))
+            .toRight(GenericError("No blocks in blockchain"))
+    )
+  }
+
+  def getValidBlockDelay(height: Int, account: KeyPair, refBlockBT: Long, balance: Long): Either[ValidationError, Long] = {
+    val pc = pos(height)
+
+    getHit(height, account)
+      .map(pc.calculateDelay(_, refBlockBT, balance))
   }
 
-  def getValidBlockDelay(height: Int, accountPublicKey: PublicKey, refBlockBT: Long, balance: Long): Either[ValidationError, Long] =
-    getHit(height, accountPublicKey)
-      .map(pos(height).calculateDelay(_, refBlockBT, balance))
-
-  def validateBlockDelay(height: Int, block: Block, parent: BlockHeader, effectiveBalance: Long): Either[ValidationError, Unit] = {
-    getValidBlockDelay(height, block.header.generator, parent.baseTarget, effectiveBalance)
-      .map(_ + parent.timestamp)
-      .ensureOr { mvt =>
-        GenericError(s"Block timestamp ${block.header.timestamp} less than min valid timestamp $mvt")
-      }(ts => ts <= block.header.timestamp)
-      .map(_ => ())
+  def validateBlockDelay(parentHeight: Int, block: Block, parent: BlockHeader, effectiveBalance: Long): Either[ValidationError, Unit] = {
+    for {
+      parentHitSource <- getHitSource(parentHeight)
+      gs <- if (vrfActivated(parentHeight + 1)) {
+        crypto.verifyVRF(block.header.generationSignature, parentHitSource, block.header.generator).map(_.arr)
+      } else {
+        generationSignature(parentHitSource, block.header.generator).asRight[ValidationError]
+      }
+      ts = pos(parentHeight).calculateDelay(hit(gs), parent.baseTarget, effectiveBalance) + parent.timestamp
+      _ <- Either.cond(
+        ts <= block.header.timestamp,
+        (),
+        GenericError(s"Block timestamp ${block.header.timestamp} less than min valid timestamp $ts")
+      )
+    } yield ()
   }
 
   def validateGenerationSignature(block: Block): Either[ValidationError, ByteStr] = {
-    val h             = blockchain.height
-    val prevHitSource = blockchain.hitSourceAtHeight(h).toRight(GenericError(s"Couldn't find hit source at height: $h"))
-    val genSig        = block.header.generationSignature
-    val generator     = block.header.generator
-
-    if (vrfActivated(h + 1))
-      prevHitSource.flatMap(crypto.verifyVRF(genSig, _, generator))
-    else
-      prevHitSource
-        .map(phs => ByteStr(generationSignature(phs.arr, generator)))
+    val blockGenSig = block.header.generationSignature.arr
+    val height      = blockchain.height
+
+    if (vrfActivated(height + 1)) {
+      getHitSource(height).flatMap(hs => crypto.verifyVRF(blockGenSig, hs, block.header.generator))
```

Comment:
Please change an error message in `verifyVRF` to something more generic like "could not verify VRF proof" and also remove logging.